stablehlo/transforms/StablehloRefineShapes.cpp

```diff
@@ -516,27 +516,75 @@ LogicalResult refineReturnTypes(PatternRewriter& rewriter, Operation* op,
     });
 
   SmallVector<Type> flattenedRefinedTypes;
-  for (auto [currentType, refinement] :
-       llvm::zip(flattenedTypes, refinements)) {
-    auto refinedElementType = refinement.getElementType();
-    if (!refinedElementType) {
-      auto currentShapedType = currentType.dyn_cast<ShapedType>();
-      if (!currentShapedType)
-        return rewriter.notifyMatchFailure(
-            op, "refineReturnTypes failed: expected shaped return types");
-      refinedElementType = currentShapedType.getElementType();
+  for (auto it : llvm::zip(flattenedTypes, refinements)) {
+    // Cannot use structured bindings to simplify this because capturing
+    // structured bindings in a lambda is a C++ 20 extension.
+    ShapedType currentType = std::get<0>(it).dyn_cast<ShapedType>();
+    ShapedTypeComponents refinement = std::get<1>(it);
+    auto failure = [&](StringRef reason) {
```

Comment:
Can we pick a different name to avoid confusion with the typical `return failure();`

Reply:
`failWithDiagnostic` perhaps

Reply:
Done.